### PR TITLE
Immutability for data collators

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?
Introduces new tests that check if a data collator might introduce side effects, i.e. the given input changes after the call to the collator. Motivated by #30556 

Furthermore, fixes the seq2seq collator to not introduce side effects on the given input's labels. This is done by:
- Passing only relevant features to the tokenizer to pad.
- Manually crafting the labels afterwards.
- Reintroducing tokenizer behaviour by converting labels to the respective datatype (pt, tf, np).
As a side note, added some more checks on `None` labels, especially when given `"labels": None` in the dictionary. 

Last remarks: 
- The test handles most of the cases that are introduced in the base tests but if I missed something give me a heads up :D 
- I'm not sure how to handle it when the user introduces `None` labels. For now, I return them by `None` again.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts @Rocketknight1

### CODE REVIEW

**Comment by vasqu** [comment]
Oh yea, one last thing. I've created separated classes for the immutability tests. Thought it got too convoluted otherwise.

File: src/transformers/data/data_collator.py [hunk 1]
```diff
@@ -585,51 +585,85 @@ class DataCollatorForSeq2Seq:
...
-                    feature["labels"] = np.concatenate([feature["labels"], remainder]).astype(np.int64)
-                else:
-                    feature["labels"] = np.concatenate([remainder, feature["labels"]]).astype(np.int64)
 
-        features = pad_without_fast_tokenizer_warning(
+        label_name = "label" if "label" in features[0].keys() else "labels"
+        labels = [feature[label_name] for feature in features] if label_name in features[0].keys() else None
+        # reconvert list[None] to None if necessary
+        # this might occur when we pass {..., "labels": None}
+        if labels is not None and all(label is None for label in labels):
```
Comment (vasqu) [inline src/transformers/data/data_collator.py:593]: Potentially change `all` to `any`. Not sure when and if someone would do this, but you never know.

File: src/transformers/data/data_collator.py [hunk 2]
```diff
@@ -585,51 +585,85 @@ class DataCollatorForSeq2Seq:
...
+                batch["labels"] = torch.tensor(batch["labels"], dtype=torch.int64)
+            elif return_tensors == "tf":
+                import tensorflow as tf
+
+                batch["labels"] = tf.constant(batch["labels"], dtype=tf.int64)
+            else:
+                batch["labels"] = np.array(batch["labels"], dtype=np.int64)
+        else:
+            # TODO need confirmation how this should be handled
+            batch["labels"] = None
```
Comment (Rocketknight1) [inline src/transformers/data/data_collator.py:655]: I think if the input `labels` is `None` then the output should be as well, so this is fine!

**Review by Rocketknight1 — APPROVED** [review]
This looks good to me - I'm happy with the change to the data collator code itself! The one issue I'd raise is that it seems like there's a lot of code duplication in the tests, with just some changes like `torch.tensor` -> `np.array` or different values for `return_tensors`. This is probably fine, though - that kind of explicit duplication makes it easier to locate test errors.

cc @amyeroberts for core maintainer review (and also let me know if you think I'm wrong about the code duplication)

**Comment by vasqu** [comment]
@Rocketknight1 With code duplications, do you mean across the different classes between pt/tf/np? I agree with that, it's more so a dependency check to see that it suddenly doesn't branch out to something unwanted and that different input datatypes are handled correctly. For example, when the default collator is used, it branches into separate np, tf, and pt calls (i.e. `numpy_default_data_collator`, `tf_default_data_collator`, `torch_default_data_collator`). I haven't deep-dived where else something like this might happen.

**Comment by Rocketknight1** [comment]
Yes, that's what I was referring to - and I think it's fine to keep it as-is!

File: tests/trainer/test_data_collator.py [hunk 3]
```diff
@@ -439,6 +439,289 @@ def test_sop(self):
...
+
+        for datatype_input, datatype_label in [
+            (list, int),
+            (list, float),
+            (np.array, int),
+            (np.array, torch.tensor),
+            (list, self._turn_to_none),
+        ]:
+            self._compare_assert_with_collator_on_datatypes(
+                default_data_collator, features_base_single_label, "inputs", datatype_input, "label", datatype_label
```
Comment (amyeroberts) [inline tests/trainer/test_data_collator.py:507]: To make the test clearer - let's call the test with kwargs rather than args

**Review by amyeroberts — COMMENTED** [review]
Thanks for tackling this and adding tests ❤️ 

Just a few formatting style comments (to be propogated to TF and Flax testers too). Otherwise looks great

File: tests/trainer/test_data_collator.py [hunk 4]
```diff
@@ -439,6 +439,289 @@ def test_sop(self):
...
+        # we go through every item and convert to `primitive` datatypes if necessary
+        # then compares for equivalence for the original data and the data that has been passed through the collator
+        for i in range(len(original)):
+            for key in original[i].keys():
+                if isinstance(original[i][key], np.ndarray):
+                    self.assertEqual(original[i][key].tolist(), batch[i][key].tolist())
+                elif isinstance(original[i][key], torch.Tensor):
+                    self.assertEqual(original[i][key].tolist(), batch[i][key].tolist())
+                else:
+                    self.assertEqual(original[i][key], batch[i][key])
```
[the remark this answers is not in the document]
Comment (vasqu) [reply tests/trainer/test_data_collator.py]: Yea, the datatype of `original_data` and `batch_data` is `List[Dict[Any]]` where any usually is a list or tensor etc.

Have changed the code to iterate via zip.

File: tests/trainer/test_data_collator.py [hunk 5]
```diff
@@ -439,6 +439,289 @@ def test_sop(self):
...
+            vocab_writer.write("".join([x + "\n" for x in vocab_tokens]))
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpdirname)
+
+    def _turn_to_none(self, item):
+        """used to convert `item` to `None` type"""
+        return None
+
+    def _compare_assert_with_collator(self, collator, original, batch):
```
[the remark this answers is not in the document]
Comment (vasqu) [reply tests/trainer/test_data_collator.py]: Renamed to `_validate_original_data_against_collated_data` but open to other suggestions.

File: tests/trainer/test_data_collator.py [hunk 6]
```diff
@@ -439,6 +439,289 @@ def test_sop(self):
...
+        for i in range(len(original)):
+            for key in original[i].keys():
+                if isinstance(original[i][key], np.ndarray):
+                    self.assertEqual(original[i][key].tolist(), batch[i][key].tolist())
+                elif isinstance(original[i][key], torch.Tensor):
+                    self.assertEqual(original[i][key].tolist(), batch[i][key].tolist())
+                else:
+                    self.assertEqual(original[i][key], batch[i][key])
+
+    def _compare_assert_with_collator_on_datatypes(
```
[the remark this answers is not in the document]
Comment (vasqu) [reply tests/trainer/test_data_collator.py]: Renamed to `_validate_original_data_against_collated_data_on_specified_keys_and_datatypes`. Might be a bit long tho.

File: tests/trainer/test_data_collator.py (hunk 3 above)
Comment (vasqu) [reply tests/trainer/test_data_collator.py]: Propagated to other similar calls.

**Review by amyeroberts — APPROVED** [review]
Thanks for all the work adding this and iterating!
